src/fs.rs

```diff
@@ -1,278 +1,955 @@
+        let output = Arc::new(AtomicCell::new(Ok(())));
+        let err = io::Error::last_os_error();
+        let err2 = io::Error::last_os_error();
+        local.push(MockWork::Err(err));
+        pull_work(
+            &local,
+            global,
+            Arc::new(Vec::new()),
+            waker,
+            finished.clone(),
+            output.clone(),
+            1,
+        );
+        assert_eq!(err2.kind(), unwrap_output(output).unwrap_err().kind());
+        assert_eq!(true, finished.load(Relaxed));
+    }
+
+    #[test]
+    fn pull_ok() {
+        let local = Worker::new_lifo();
+        let global = Arc::new(Injector::new());
+        let waker = Arc::new((Mutex::new(1), Condvar::new()));
+        let finished = Arc::new(AtomicBool::new(true));
+        let output = Arc::new(AtomicCell::new(Ok(())));
+        local.push(MockWork::Ok);
+        pull_work(
+            &local,
+            global,
+            Arc::new(Vec::new()),
+            waker,
+            finished,
+            output.clone(),
+            1,
+        );
+        assert_eq!((), unwrap_output(output).unwrap());
+        // the unparker should trigger the condvar, so that errors cascade across all the
+        // threads rapidly even if many are waiting; but again this is tricky to
+        // test :/.
+    }
+
+    #[test]
+    fn pull_empty() {
+        let local: Worker<MockWork> = Worker::new_lifo();
+        let global = Arc::new(Injector::new());
+        let waker = Arc::new((Mutex::new(0), Condvar::new()));
+        let finished = Arc::new(AtomicBool::new(true));
+        let output = Arc::new(AtomicCell::new(Ok(())));
+        let child_waker = waker.clone();
+        let child_output = output.clone();
+        let child = thread::spawn(move || {
+            // Tell the test thread we've started.
+            {
+                let mut guard = child_waker.0.lock().unwrap();
+                // Tells the outer thread to continue, and also tells
+                // wait_for_threads that we've completed.
+                *guard = 1;
+                child_waker.1.notify_one();
+            }
+            pull_work(
+                &local,
+                global,
+                Arc::new(Vec::new()),
+                child_waker,
+                finished,
+                child_output,
+                1,
+            )
+        });
+
+        {
+            // wait for the test thread to start
+            let mut guard = waker.0.lock().unwrap();
+            while *guard != 1 {
+                guard = waker.1.wait(guard).unwrap();
             }
         }
+        thread::sleep(Duration::from_millis(50));
+        // wake the child
+        waker.1.notify_one();
+        child.join().unwrap();
+
+        assert_eq!((), unwrap_output(output).unwrap());
     }
 }
+
+// tests to write:  scandir empty should immediately deleteDir
+//                  deletedir no parent should ???
+//                  wait_for_threads
```

Comment:
If this is future work, could this be a issue rather than at the end of file?

Reply:
It'll be done before merging